docs/en/config.md
# Configuration

The application configuration is mainly placed into the `framarama/settings.py`
file. Please adjust settings there.

### General

#### `SECRET_KEY`

Default: temporary key

This configures a secret key which is used internally for different things. Please
do not use the default and change it to something really serect.

#### `FRAMARAMA.MODES`

Default: `server` and `frontend`

Defines which component to start: server, frontend or both. Just comment in or out
to define what to start.

Only start server component:

```
'MODES': [
    'server',
],
```

Only start frontend component:

```
'MODES': [
    'frontend',
],
```

Start both components:

```
'MODES': [
    'server',
    'frontend',
],
```

#### `FRAMARAMA.DATA_PATH`

Default: `./data/`

Should point to a directory where the application can store some data. Please
do not point this to a temporary directory. It will contain persistent data
which should not be deleted.

#### `FRAMARAMA.MEDIA_PATH`

Default: `./data/media/`

Points to a directory where some media files can be placed to read. The directory
source plugin will read the files located in this directory.

#### `FRAMARAMA.MOUNT_PATH`

Default: `/media`

To automatically provide mounted storaged in media path to use for picture collecton,
this settings points to the standard system mount path for external storages (e.g.
when using `automount`).

#### `FRAMARAMA.MOUNT_LEVEL`

Default: `1`

When `MOUNT_PATH` is configured to watch for new mounted storaged, this setting
will define how many directory levels to check (useful when external storages
will be mounted in sub-directory of `MOUNT_PATH` (e.g. `/media/username/ext-storage`).
Using level of `0` will make directories one level below `MOUNT_PATH` available,
`1` will make directories two levels below `MOUNT_PATH` as media directories (e.g.
`ext-storage`).


### Server

#### `FRAMARAMA.ADMIN_USERNAME`

Default: `admin`

Contains the username for the adminstrative user which will be created initially
when starting the server application for the first time.

#### `FRAMARAMA.ADMIN_PASSWORD`

Default: `testabc123`

The simple administrator password - please change it!

#### `FRAMARAMA.ADMIN_MAIL`

Default: `admin@some-domain.tld`

This is the mail address used for the administrative user. Currently it does not
have any effect, and do not require a valid mail address.

#### `FRAMARAMA.CONFIG_THUMBNAIL_SIZE`

Default: `[640, 480]`

When the frontend component submits the photo to the server it is also providing
a thumbnail of the current photo. If not specified the thumbnail is generated
on the server side.

#### `FRAMARAMA.CONFIG_SOURCE_UPDATE_INTERVAL`

Default: `23:00:00`

The interval (hours, minutes, seconds) when the server is updating the
photo collection and adding, removing or updating them on the server. Set the
value to `None` to disable automatic updates completely.

### Frontend

#### `FRAMARAMA.AP_NAME`

Default: `framaRAMA`

In case the wireless network connection can not be established a fallback
access point will be started with this name.

#### `FRAMARAMA.AP_PASS`

Default: `framarama`

When the fallback access point is started it is using this password.

#### `FRAMARAMA.FRONTEND_KEYSTROKES`

Default: `True`

Register key strokes to control the frontend device when connecting a keyboard
to it (e.g. activate wireless access point, reboot device, etc.).

#### `FRAMARAMA.FRONTEND_ITEM_UPDATE_INTERVAL`

Default: `00:05:00`

The inverval (hours, minutes, seconds) when the frontend is updating the
display settings and applying them.

#### `FRAMARAMA.FRONTEND_THUMBNAIL_SIZE`

Default: `[640, 480]`

This defines the resolution to use when generating a thumbnail of the
photos displayed on the frontend. It is used in the frontend application
and - when activated - when sumbmitting them to the server.

#### `FRAMARAMA.FRONTEND_APP_UPDATE_INTERVAL`

Default: `23:00:00`

Default the interval for the update check. This will only fetch the updates
but does not install them.

#### `FRAMARAMA.FRONTEND_APP_UPDATE_PRECMD`

Default: ``

This command is use in every case the software is updated. It can be used to
prepare something before running the update (e.g. mount something, delete
something, etc.).

#### `FRAMARAMA.FRONTEND_APP_UPDATE_POSTCMD`

Default: ``

Like the pre-command this is a command executed after the software update
is completed.

### Docker

When using the docker container the following environment variables can
be used. Some of them refer to the settings listed above, others are docker-only
settings.

#### `DJANGO_DEUBG`

Default: `True`

Enables or disable Django debug logging and debug mode.

#### `DJANGO_SECRET_KEY`

Default: some non-secure secret

Configures the common Django secret key. This should be set in any environment to
a secure value.

#### `FRAMARAMA_MODES`

See `FRAMARAMA.MODES`.

#### `FRAMARAMA_PORT`

Default: `8000`

Specifies the HTTP port to use for the application.

#### `FRAMARAMA_APP_UPDATE_PRECMD`

Default: (empty)

See `FRAMARAMA.FRONTEND_APP_UPDATE_PRECMD`.

#### `FRAMARAMA_APP_UPDATE_POSTCMD`

Default: (empty)

See `FRAMARAMA.FRONTEND_APP_UPDATE_POSTCMD`.



requirements/config.txt
-r common.txt
#mysqlclient==2.1.0
Wand==0.6.7



framarama/settings.py
"""
Django settings for framarama project.

Generated by 'django-admin startproject' using Django 4.0.4.

For more information on this file, see
https://docs.djangoproject.com/en/4.0/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/4.0/ref/settings/
"""

from os import path, getenv, environ
from pathlib import Path

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/4.0/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = environ.get('DJANGO_SECRET', 'django-insecure-@+o7lclaqfg8rkb0dp$q&eibn1b$-#-!)@ayalffvx7ztlwkf@')

# SECURITY WARNING: don't run with debug turned on in production!
# But required when serving static files via "runserver" command!
DEBUG = environ.get('DJANGO_DEBUG', 'True') == 'True'

ALLOWED_HOSTS = ['*']


# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'rest_framework',
    'account',
    'common',
    'config',
    'api',
    'frontend',
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'framarama.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.jinja2.Jinja2',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'environment': 'framarama.jinja.environment',
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        }
    },
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'framarama.wsgi.application'

# Logging
# https://docs.djangoproject.com/en/4.0/topics/logging/#configuring-logging
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'console': {
            'class': 'logging.StreamHandler',
        },
    },
    'loggers': {
        'django.db.backends': {
            'level': 'DEBUG' if DEBUG else 'INFO',
            'handlers': ['console'],
            'propagate': False,
        },
        'framarama': {
            'level': 'INFO',
            'handlers': ['console'],
            'propagate': False,
        },
    },
    'root': {
        'handlers': ['console'],
        'level': 'DEBUG' if DEBUG else 'INFO',
    },
}

# Database
# https://docs.djangoproject.com/en/4.0/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': BASE_DIR / 'data' / 'db.sqlite3',
    },
}
DATABASE_ROUTERS = [
    'framarama.base.DatabaseRouter'
]
def configure_databases(reload=True):
    if path.exists(BASE_DIR / 'framarama' / 'settings_db.py'):
        from django.db import connections
        from framarama import settings_db
        if reload:
            import importlib
            from framarama import base
            importlib.reload(settings_db)
            if hasattr(settings_db, 'dbs'):
                DATABASES.update(settings_db.dbs)
                connections.close_all()
                connections.settings = DATABASES
                for name in DATABASES:
                    if hasattr(connections._connections, name):
                        delattr(connections._connections, name)
            base.DatabaseRouter.config = None     # use new settings in database router
        elif hasattr(settings_db, 'dbs'):
            DATABASES.update(settings_db.dbs)

configure_databases(False)

# Password validation
# https://docs.djangoproject.com/en/4.0/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

AUTH_USER_MODEL = 'account.User'

# Internationalization
# https://docs.djangoproject.com/en/4.0/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/4.0/howto/static-files/

STATIC_URL = 'static/'
STATIC_ROOT = BASE_DIR / 'static'
STATICFILES_FINDERS = [
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
]

# Default primary key field type
# https://docs.djangoproject.com/en/4.0/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'

# Configure login
LOGIN_URL = '/config/accounts/login'
LOGIN_REDIRECT_URL = '/config/'
LOGOUT_REDIRECT_URL = LOGIN_URL

REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': [
        'api.auth.TokenAuthentication',
        'rest_framework.authentication.BasicAuthentication',
        'rest_framework.authentication.SessionAuthentication',  # required for API browser
    ],
    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.LimitOffsetPagination',
    'PAGE_SIZE': 100,
}

FRAMARAMA = {
    'TITLE': 'framaRAMA',
    'MODES': environ.get('FRAMARAMA_MODES', 'server,frontend').split(','),
    'DATA_PATH': './data/',
    'MEDIA_PATH': './data/media/',
    'MOUNT_PATH': '/media/',
    'MOUNT_LEVEL': 1,
    'COMMON_PATH': './common/static/common/',
    'ADMIN_USERNAME': 'admin',
    'ADMIN_PASSWORD': 'testabc123',
    'ADMIN_MAIL': 'admin@some-domain.tld',
    'API_URL': 'http://127.0.0.1:' + environ.get('FRAMARAMA_PORT', '8000'),
    'IMAGE_PROCESSING_ADAPTER': 'config.utils.finishing.WandImageProcessingAdapter',
    'GIT_REMOTE': 'origin',
    'AP_NAME': 'framaRAMA',
    'AP_PASS': 'framarama',
    'FRONTEND_KEYSTROKES': False,
    'FRONTEND_ITEM_UPDATE_INTERVAL': '00:05:00',
    'FRONTEND_THUMBNAIL_SIZE': [640, 480],
    'FRONTEND_APP_UPDATE_INTERVAL': '23:00:00',
    'FRONTEND_APP_UPDATE_PRECMD': environ.get('FRAMARAMA_APP_UPDATE_PRECMD', ''),
    'FRONTEND_APP_UPDATE_POSTCMD': environ.get('FRAMARAMA_APP_UPDATE_POSTCMD', ''),
    'CONFIG_THUMBNAIL_SIZE': [640, 480],
    'CONFIG_SOURCE_UPDATE_INTERVAL': '23:00:00',
    'CONFIG_SORTING_EVAL_QUERY': False,
}

